File: content/snippets/typescript/interview_lcci/17/07_baby_names.ts
function trulyMostPopular(names: string[], synonyms: string[]): string[] {
    const parent = new Map<string, string>();

    const find = (x: string): string => {
        if (!parent.has(x)) parent.set(x, x);
        if (parent.get(x)! !== x) parent.set(x, find(parent.get(x)!));
        return parent.get(x)!;
    };

    const unite = (a: string, b: string): void => {
        const pa = find(a);
        const pb = find(b);
        if (pa !== pb) {
            if (pa < pb) parent.set(pb, pa);
            else parent.set(pa, pb);
        }
    };

    const freq = new Map<string, number>();
    for (const s of names) {
        const idx = s.indexOf('(');
        const name = s.slice(0, idx);
        const count = Number(s.slice(idx + 1, -1));
        freq.set(name, count);
    }

    for (const syn of synonyms) {
        const s = syn.slice(1, -1);
        const comma = s.indexOf(',');
        unite(s.slice(0, comma), s.slice(comma + 1));
    }

    const result = new Map<string, number>();
    for (const [name, cnt] of freq) {
        const root = find(name);
        result.set(root, (result.get(root) ?? 0) + cnt);
    }

    return [...result.entries()]
        .sort((a, b) => a[0].localeCompare(b[0]))
        .map(([name, cnt]) => `${name}(${cnt})`);
}

const names = ["John(15)", "Jon(12)", "Chris(13)", "Kris(4)", "Christopher(19)"];
const synonyms = ["(Jon,John)", "(John,Johnny)", "(Chris,Kris)", "(Chris,Christopher)"];
console.log(trulyMostPopular(names, synonyms).join(' ')); // Chris(36) John(27)

File: content/snippets/scala2/interview_lcci/17/07_baby_names.scala
object Lcci1707BabyNames {
  def trulyMostPopular(names: Array[String], synonyms: Array[String]): Array[String] = {
    val parent = scala.collection.mutable.Map[String, String]()

    def find(x: String): String = {
      if (!parent.contains(x)) parent(x) = x
      if (parent(x) != x) parent(x) = find(parent(x))
      parent(x)
    }

    def unite(a: String, b: String): Unit = {
      val pa = find(a)
      val pb = find(b)
      if (pa != pb) {
        if (pa < pb) parent(pb) = pa
        else parent(pa) = pb
      }
    }

    val freq = scala.collection.mutable.Map[String, Int]()
    for (s <- names) {
      val idx = s.indexOf('(')
      val name = s.substring(0, idx)
      val count = s.substring(idx + 1, s.length - 1).toInt
      freq(name) = count
    }

    for (syn <- synonyms) {
      val s = syn.substring(1, syn.length - 1)
      val comma = s.indexOf(',')
      unite(s.substring(0, comma), s.substring(comma + 1))
    }

    val result = scala.collection.mutable.Map[String, Int]().withDefaultValue(0)
    for ((name, cnt) <- freq) result(find(name)) += cnt

    result.toArray.sortBy(_._1).map { case (name, cnt) => s"$name($cnt)" }
  }

  def main(args: Array[String]): Unit = {
    val names = Array("John(15)", "Jon(12)", "Chris(13)", "Kris(4)", "Christopher(19)")
    val synonyms = Array("(Jon,John)", "(John,Johnny)", "(Chris,Kris)", "(Chris,Christopher)")
    println(trulyMostPopular(names, synonyms).mkString(" ")) // Chris(36) John(27)
  }
}
